src/eos-updater-poll.c

```diff
@@ -153,65 +308,181 @@ metadata_fetch_finished (GObject *object,
 }
 
 static void
+maybe_send_metric (EosMetricsInfo *metrics)
+{
+#ifdef HAS_EOSMETRICS_0
+  static gboolean metric_sent = FALSE;
+
+  if (metric_sent)
+    return;
+
+  message ("Recording metric event %s: (%s, %s, %s, %d)",
+           EOS_UPDATER_BRANCH_SELECTED, metrics->vendor, metrics->product,
+           metrics->ref, metrics->on_hold);
+  emtr_event_recorder_record_event_sync (emtr_event_recorder_get_default (),
+                                         EOS_UPDATER_BRANCH_SELECTED,
+                                         g_variant_new ("(sssb)", metrics->vendor,
+                                                        metrics->product,
+                                                        metrics->ref,
+                                                        metrics->on_hold));
+  metric_sent = TRUE;
+#endif
+}
+
+typedef struct UpdateAndMetrics UpdateAndMetrics;
+
+struct UpdateAndMetrics
+{
+  EosUpdateInfo *update;
+  EosMetricsInfo *metrics;
+};
+
+static UpdateAndMetrics *
+update_and_metrics_new (EosUpdateInfo *update,
+                        EosMetricsInfo *metrics)
+{
+  UpdateAndMetrics *uam = g_new0 (UpdateAndMetrics, 1);
+
+  if (update != NULL)
+    uam->update = g_object_ref (update);
+
+  if (metrics != NULL)
+    uam->metrics = g_object_ref (metrics);
+
+  return uam;
+}
+
+static void
+update_and_metrics_free (UpdateAndMetrics *uam)
+{
+  if (uam == NULL)
+    return;
+
+  g_clear_object (&uam->update);
+  g_clear_object (&uam->metrics);
+  g_free (uam);
+}
+
+static UpdateAndMetrics *
+get_latest_uam (EosUpdaterData *data,
+                GHashTable *source_to_uam,
+                gboolean with_updates)
+{
+  g_autoptr(GHashTable) latest = g_hash_table_new (NULL, NULL);
+  GHashTableIter iter;
+  gpointer name_ptr;
+  gpointer uam_ptr;
+  GDateTime *latest_timestamp = NULL;
+  gsize idx;
+
+  g_hash_table_iter_init (&iter, source_to_uam);
+  while (g_hash_table_iter_next (&iter, &name_ptr, &uam_ptr))
+    {
+      UpdateAndMetrics *uam = uam_ptr;
+      EosBranchFile *branch_file = uam->metrics->branch_file;
+
+      if (with_updates && uam->update == NULL)
+        continue;
+
+      if (latest_timestamp != NULL)
+        {
+          if (g_date_time_compare (branch_file->download_time,
+                                   latest_timestamp))
+            {
+              latest_timestamp = NULL;
+              g_hash_table_remove_all (latest);
+            }
+        }
+      if (latest_timestamp == NULL)
+        {
+          latest_timestamp = branch_file->download_time;
+          g_hash_table_insert (latest, name_ptr, uam_ptr);
```

Comment:
Meh, you just found a bug. :P

The `name_ptr` and `uam_ptr` should be added to the `latest` hash table also when timestamp is the same as the latest one. This could happen if we got branch files with the same timestamp from both main server and some LAN server.

The hash table is used below in the `for` loop to get the `uam` from the preferred source.